File: src/shader.c
#include "shader.h"

#include <GL/glew.h>
#include <GL/glx.h>
#include <GL/gl.h>

#include <stdio.h>
#include <stdlib.h>
#include <string.h>

char* readFile(char *filepath) {
	FILE *file;
	char *buffer;
	long lSize;

	/* open file read only */
	if (!(file = fopen(filepath, "r"))) {
		fprintf(stderr, "Failed to open file\n");
		return NULL;
	}

	fseek(file, 0L, SEEK_END);
	lSize = ftell(file);
	rewind(file);

	/* allocate memory for entire content */
	buffer = (char*) calloc( 1, lSize+1 );
	if (!buffer) {
		fclose(file);
		fprintf(stderr, "Memory alloc fails\n");
		return NULL;
	}

	/* copy the file into the buffer */
	if (fread( buffer , lSize, 1 , file) != 1) {
		fclose(file);
		free(buffer);
		fprintf(stderr, "Entire read fails\n");
		return NULL;
	}

	fclose(file);

	return buffer;
}

int shader_check_compile_errors(Shader shader, const char *type) {
	char infoLog[1024];
	int success;

	if (strcmp(type, "PROGRAM") != 0) {
		glGetShaderiv(shader, GL_COMPILE_STATUS, &success);
		if (!success) {
			glGetShaderInfoLog(shader, 1024, NULL, infoLog);
			fprintf(stderr, "ERROR::SHADER_COMPILATION_ERROR of type: %s\n%s\n", type, infoLog);
			return 0;
		}
	} else {
		glGetProgramiv(shader, GL_LINK_STATUS, &success);
		if (!success) {
			glGetProgramInfoLog(shader, 1024, NULL, infoLog);
			fprintf(stderr, "ERROR::PROGRAM_LINKING_ERROR of type: %s\n%s\n", type, infoLog);
			return 0;
		}
	}
	return 1;
}

Shader shader_compile(char* vShaderCode, char* fShaderCode) {
	unsigned int vertex, fragment;
	Shader shader = 0;

	/* vertex shader */
	vertex = glCreateShader(GL_VERTEX_SHADER);
	glShaderSource(vertex, 1, (const GLchar *const *)&vShaderCode, NULL);
	glCompileShader(vertex);

	if (!shader_check_compile_errors(vertex, "VERTEX")) {
		return 0;
	}

	/* fragment Shader */
	fragment = glCreateShader(GL_FRAGMENT_SHADER);
	glShaderSource(fragment, 1, (const GLchar *const *)&fShaderCode, NULL);
	glCompileShader(fragment);

	if (!shader_check_compile_errors(fragment, "FRAGMENT")) {
		glDeleteShader(vertex);
		return 0;
	}

	/* shader Program */
	shader = glCreateProgram();
	glAttachShader(shader, vertex);
	glAttachShader(shader, fragment);
	glLinkProgram(shader);

	if (!shader_check_compile_errors(shader, "PROGRAM")) {
		glDeleteShader(vertex);
		glDeleteShader(fragment);
		return 0;
	}

	/* delete the shaders as they're linked into our program now and no longer necessary */
	glDeleteShader(vertex);
	glDeleteShader(fragment);

	return shader;
}

void shader_bind(Shader shader) {
	glUseProgram(shader);
}

void shader_unbind() {
	glUseProgram(0);
}

void shader_set_float(const int loc, float value) {
	glUniform1f(loc, value);
}

void shader_set_vec2(const int loc, float value1, float value2) {
	glUniform2f(loc, value1, value2);
}

int shader_get_location(Shader shader, const char *name) {
	return glGetUniformLocation(shader, name);
}

File: show/shader.py
from OpenGL import GL as gl

import logging
import sys

log = logging.getLogger(__name__)

class Shader():
    def __init__(self, shaders):
        log.debug('creating the shader program')
        self.program_id = gl.glCreateProgram()
        self.shader_ids = []

        for shader_type, shader_src in shaders.items():
            shader_id = gl.glCreateShader(shader_type)
            gl.glShaderSource(shader_id, shader_src)

            log.debug(f'compiling the {shader_type} shader')
            gl.glCompileShader(shader_id)

            # check if compilation was successful
            result = gl.glGetShaderiv(shader_id, gl.GL_COMPILE_STATUS)
            info_log_len = gl.glGetShaderiv(shader_id, gl.GL_INFO_LOG_LENGTH)
            if info_log_len:
                logmsg = gl.glGetShaderInfoLog(shader_id)
                log.error(logmsg)
                sys.exit(10)

            gl.glAttachShader(self.program_id, shader_id)
            self.shader_ids.append(shader_id)

        log.debug('linking shader program')
        gl.glLinkProgram(self.program_id)

        # check if linking was successful
        success = gl.glGetProgramiv(self.program_id, gl.GL_LINK_STATUS)
        info_log_len = gl.glGetProgramiv(self.program_id, gl.GL_INFO_LOG_LENGTH)
        if not success:
            logmsg = gl.glGetProgramInfoLog(self.program_id)
            log.error(logmsg)
            sys.exit(0)

        log.debug('installing shader program into rendering state')
        gl.glUseProgram(self.program_id)

    def bind(self):
        gl.glUseProgram(self.program_id)

    def get_uniform(self, name):
        return gl.glGetUniformLocation(self.program_id, name)

    def __del__(self):
        log.debug('cleaning up shader program')
        for shader_id in self.shader_ids:
            gl.glDetachShader(self.program_id, shader_id)
            gl.glDeleteShader(shader_id)
        gl.glDeleteProgram(self.program_id)
        gl.glUseProgram(0)
